lib/tests/test_conflicts.rs

```diff
@@ -1810,6 +1816,328 @@ fn test_update_conflict_from_content_simplified_conflict() {
+    <<<<<<<<<<< Conflict 1 of 1
+    +++++++++++ Contents of side #1
+    <<<< left 1
+    ----------- Contents of base
+    line 1
+    +++++++++++ Contents of side #2
+    >>>>>>> right 1
+    >>>>>>>>>>> Conflict 1 of 1 ends
+    line 2
+    line 3
+    "##
+    );
+}
+
+#[test]
+fn test_update_from_content_malformed_conflict() {
+    let test_repo = TestRepo::init();
+    let store = test_repo.repo.store();
+
+    // Create conflicts which contain conflict markers of varying lengths
+    let path = RepoPath::from_internal_string("dir/file");
+    let base_file_id = testutils::write_file(
+        store,
+        path,
+        indoc! {"
+            line 1
+            line 2
+            line 3
+            line 4
+            line 5
+        "},
+    );
+    let left_file_id = testutils::write_file(
+        store,
+        path,
+        indoc! {"
+            line 1
+            line 2 left
+            line 3
+            line 4 left
+            line 5
+        "},
+    );
+    let right_file_id = testutils::write_file(
+        store,
+        path,
+        indoc! {"
+            line 1
+            line 2 right
+            line 3
+            line 4 right
+            line 5
+        "},
+    );
+    let conflict = Merge::from_removes_adds(
+        vec![Some(base_file_id.clone())],
+        vec![Some(left_file_id.clone()), Some(right_file_id.clone())],
+    );
+
+    let materialized =
+        materialize_conflict_string(store, path, &conflict, ConflictMarkerStyle::Diff);
+    insta::assert_snapshot!(materialized, @r##"
+    line 1
+    <<<<<<< Conflict 1 of 2
+    %%%%%%% Changes from base to side #1
+    -line 2
+    +line 2 left
+    +++++++ Contents of side #2
+    line 2 right
+    >>>>>>> Conflict 1 of 2 ends
+    line 3
+    <<<<<<< Conflict 2 of 2
+    %%%%%%% Changes from base to side #1
+    -line 4
+    +line 4 left
+    +++++++ Contents of side #2
+    line 4 right
+    >>>>>>> Conflict 2 of 2 ends
+    line 5
+    "##
+    );
+
+    let parse = |conflict, content| {
+        update_from_content(conflict, store, path, content, ConflictMarkerStyle::Diff)
+            .block_on()
+            .unwrap()
+    };
+    assert_eq!(parse(&conflict, materialized.as_bytes()), conflict);
+
+    // Make a change to the second conflict that causes it to become invalid
+    let new_conflict_contents = indoc! {"
+        line 1
+        <<<<<<< Conflict 1 of 2
+        %%%%%%% Changes from base to side #1
+        -line 2
+        +line 2 left
+        +++++++ Contents of side #2
+        line 2 right
+        >>>>>>> Conflict 1 of 2 ends
+        line 3
+        <<<<<<< Conflict 2 of 2
+        %%%%%%% Changes from base to side #1
+        -line 4
+        +line 4 left
+        line 4 right
+        >>>>>>> Conflict 2 of 2 ends
+        line 5
+    "};
+    // On the first snapshot, it will parse as a conflict containing conflict
+    // markers as text
+    let new_conflict = parse(&conflict, new_conflict_contents.as_bytes());
```

Comment:
Indeed. This one is scary, and would probably provide a worse UX than the current state.

We'll have to either remember the materialized marker length or parse unpaired markers into some form of conflict.